### ISSUE TITLE
Lazy loading of modules giving error

### ISSUE THREAD

**Issue by ilivestrong** [angular/angular-cli#8641]
Hi,

I have upgraded to the latest version of Angular CLI - 1.5.4 and Angular 5. I am not able to workout a simple app, which lazily load modules.  

Strangely, if I comment some code, the app gets recompiled and works perfectly. Even if I uncomment, what I commented before, it runs smoothly without any error. But if I stop npm and restart my app, I face same issue again. 

Is it something at build time? 

### Angular-CLI - 1.5.4
```
<!--
ERROR in Error: Could not resolve module src/feature-module-one/feature1.module relative to E:/Build Learn Conquer/tests/testAOTModules/src/app/app.module.ts
    at StaticSymbolResolver.getSymbolByModule (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:31884:30)
    at StaticReflector.resolveExternalReference (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:30350:62)
    at parseLazyRoute (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:28616:55)
    at listLazyRoutes (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:28578:36)
    at visitLazyRoute (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:29995:47)
    at AotCompiler.listLazyRoutes (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler\bundles\compiler.umd.js:29963:20)
    at AngularCompilerProgram.listLazyRoutes (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler-cli\src\transformers\program.js:157:30)
    at Function.NgTools_InternalApi_NG_2.listLazyRoutes (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@angular\compiler-cli\src\ngtools_api.js:44:36)
    at AngularCompilerPlugin._getLazyRoutesFromNgtools (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:247:66)
    at Promise.resolve.then.then (E:\Build Learn Conquer\tests\testAOTModules\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:538:50)
    at process._tickCallback (internal/process/next_tick.js:103:7)

webpack: Failed to compile.
-->
```

### Repro steps

* Lazily load an angular module
* ng serve

### Observed behavior
```
<!-- Normally this includes a stack trace and some more information. -->
```

**Comment by istiti** [comment]
Can you maybe reproduce here https://stackblitz.com/edit/angular-gitter-pdkcvh or give more infos about your routes ?

**Comment by ilivestrong** [comment]
routes defined in app.module.ts 
```
const appRoutes: Routes = [
    { path: "**", component: NotFound404, pathMatch: 'full' },
     { path: 'expense', loadChildren: 'src/expense/expense.module#ExpenseModule' },
     { path: 'user', loadChildren: 'src/user/user.module#UserModule' }
];
```

routes defined in featured module:
```
export const expenseRoutes: Routes = [

    { path: 'expense', component: ExpenseGroupComponent },
    { path: 'expense/:id', component: ExpenseGroupComponent },
    { path: 'expensesheet', component: ExpenseSheetComponent },
    { path: 'expensesheet/:id', component: ExpenseSheetComponent }
];
```

**Comment by ilivestrong** [comment]
@istiti  -  Also, I tried the same stuff on the link you provided, it worked without any compilation error. Not sure what I messed up on my system.

**Comment by filipesilva** [comment]
Can you try using relative paths? At the moment you're using full paths from the project root and that isn't really supported.

**Comment by ilivestrong** [comment]
@filipesilva - It was working fine with full paths from the project root. It's only after my angular upgrade to 5, things broke.

Please see the attached .rar file of my small app(_please run npm install to install all dependencies_). May be you get better idea of what is the culprit.

Thanks.


[EPManager.zip](https://github.com/angular/angular-cli/files/1507543/EPManager.zip)

**Comment by filipesilva** [comment]
I understand that the absolute paths might have been working but I don't think it was ever the intention. Did you try with relative paths?

**Comment by david-bulte** [comment]
I have the exact same problem with relative paths. 

`    {
      path: 'tele-consult',
      loadChildren: './tele-consult/tele-consult.module#TeleConsultModule',
      canActivate: [AuthGuard]
    },
`
Here's my output:

** NG Live Development Server is listening on 0.0.0.0:9020, open your browser on http://localhost:9020/ **
Date: 2018-01-19T21:27:26.270Z
Hash: 2cf405c1104615912f78
Time: 21739ms
chunk {formviewer.module} formviewer.module.chunk.js () 13.5 kB  [rendered]
chunk {inline} inline.bundle.js (inline) 5.79 kB [entry] [rendered]
chunk {main} main.bundle.js (main) 3.06 MB [initial] [rendered]
chunk {polyfills} polyfills.bundle.js (polyfills) 2.61 MB [initial] [rendered]
chunk {scripts} scripts.bundle.js (scripts) 32.8 kB [initial] [rendered]
chunk {styles} styles.bundle.js (styles) 746 kB [initial] [rendered]
chunk {twilio-chat} twilio-chat.bundle.js (twilio-chat) 1.25 MB [initial] [rendered]
chunk {twilio-video} twilio-video.bundle.js (twilio-video) 894 kB [initial] [rendered]
chunk {vendor} vendor.bundle.js (vendor) 18.3 MB [initial] [rendered]

webpack: Compiled successfully.
webpack: Compiling...
Date: 2018-01-19T21:27:37.030Z • Hash: 3a43e0922f3742383a35 • Time: 1898ms
8 unchanged chunks
chunk {main} main.bundle.js (main) 3.06 MB [initial] [rendered]
chunk {tele-consult.module} tele-consult.module.chunk.js () 1.04 MB  [rendered]

You can see that the to-be-lazy-loaded module is not created initially, but only after making a little change, thereby forcing a recompile.

I also see "WARNING in Duplicated path in loadChildren detected during a rebuild. We will take the latest version detected and override it to save rebuild time. You should perform a full build to validate that your routes don't overlap." Don't know if it's of any significance?

**Comment by Mage111** [comment]
I'm using relative paths, and also have the same issue after upgrading to angular 5

**Comment by Greatshock** [comment]
+1

**Comment by Yura13** [comment]
I have the same error after upgrading:
angular 4.3.1 => 5.2.3
cli 1.2.3 => 1.6.7
typescript 2.4.2 => 2.5.3
rxjs 5.4.2 => 5.5.6
zone.js 0.8.14 => 0.8.20
```
ERROR in Error: Could not resolve module app/admin/admin.module.ts relative to C:/Projects/mean-app/src/app/app-routing.module.ts
    at StaticSymbolResolver.getSymbolByModule (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:29765:30)
    at StaticReflector.resolveExternalReference (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:31610:62)
    at parseLazyRoute (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:29049:55)
    at listLazyRoutes (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:29011:36)
    at visitLazyRoute (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:31175:47)
    at AotCompiler.listLazyRoutes (C:\Projects\mean-app\node_modules\@angular\compiler\bundles\compiler.umd.js:31143:20)
    at AngularCompilerProgram.listLazyRoutes (C:\Projects\mean-app\node_modules\@angular\compiler-cli\src\transformers\program.js:156:30)
    at Function.NgTools_InternalApi_NG_2.listLazyRoutes (C:\Projects\mean-app\node_modules\@angular\compiler-cli\src\ngtools_api.js:44:36)
    at AngularCompilerPlugin._getLazyRoutesFromNgtools (C:\Projects\mean-app\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:248:66)
    at Promise.resolve.then.then (C:\Projects\mean-app\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:571:50)
    at process._tickCallback (internal/process/next_tick.js:109:7)
```

Here is src/app/app-routing.module.ts:
```
const routes: Routes = [
  { path: '', redirectTo: 'login', pathMatch: 'full' },
  { path: 'login', component: LoginComponent, canActivate: [LoginGuard] },
  { path: 'admin', loadChildren: 'app/admin/admin.module.ts#AdminModule', canLoad: [AdminGuard] },
  { path: '**', redirectTo: '' }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule {}
```
Build will finish successfully if delete this line:
```{ path: 'admin', loadChildren: 'app/admin/admin.module.ts#AdminModule', canLoad: [AdminGuard] }```

Also changing to relative path didn't help.

If it will be helpful, other files:

-tsconfig.json
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ],
    "paths": {
      "core": [ "app/core" ],
      "core/*": [ "app/core/*" ],
      "shared": [ "app/shared" ],
      "shared/*": [ "app/shared/*" ]
    }
  }
}
```

-tsconfig.app.json
```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "baseUrl": "./",
    "module": "es2015",
    "types": []
  },
  "exclude": [
    "test.ts",
    "**/*.spec.ts"
  ]
}
```

-main.ts
```
if (environment.production) {
  enableProdMode();
}

platformBrowserDynamic().bootstrapModule(AppModule)
  .catch(err => console.log(err));
```

**Comment by Yura13** [comment]
I found the reason of this error.
You should change all lazyload modules in your application like this example(remove ```.ts```):
```admin.module.ts#AdminModule => admin.module#AdminModule```
If you change only one parent lazyload module (but not syntax of child lazyload modules), you have the same error related to parent module:
```Could not resolve module app/admin/admin.module.ts relative to C:/Projects/mean-app/src/app/app-routing.module.ts```

**Comment by avasarala786** [comment]
[https://stackoverflow.com/questions/47601637/angular-4-could-not-resolve-submodule-for-for-routing]
We have a solution here

**Comment by scunningham777** [comment]
Seeing this same issue on a dead simple app with Angular 6.0.5 and Angular CLI 6.0.8 (used CLI to generate a single feature module with routing and two components within that module, then set up lazy-load routing in app.module.ts). 

Changing from an absolute to a relative path fixed the issue for me; if this is the only supported approach, someone should update the [official Angular docs](https://angular.io/guide/router#lazy-loading-route-configuration), which definitely use absolute paths in the examples!

**Comment by AmebaBrain** [comment]
One obvious addition, relative path (to app-routing.module.ts) should start with dot symbol
`app/admin/admin.module#AdminModule => ./admin/admin.module#AdminModule`
Such form doesn't work
`app/admin/admin.module#AdminModule => admin/admin.module#AdminModule`

**Comment by abhi-das** [comment]
@AmebaBrain Thank you.

Working on "@angular/core": "^6.0.3" and had the same issue as mentioned above everyone.

You just need to change 
'./app/admin/...' => './admin/admin....'

as mentioned @AmebaBrain 
app/admin/admin.module#AdminModule => ./admin/admin.module#AdminModule

**Comment by adumesny** [comment]
the current angular 6 doc https://angular.io/guide/lazy-loading-ngmodules#configure-the-routes has the same issue still, and you get the error as well.

need to change:
app/customers/customers.module#CustomersModule => ./customers/customers.module#CustomersModule

**Comment by mariamtsartsidze** [comment]
I use relative paths and start with dot symbol but still have got the same problem. Any solution?

**Comment by montella1507** [comment]
@filipesilva  i think there is a bug at all...  Please download that sample - it is almost empty project just to show angular to collegues:
[build-lazy-bug.zip](https://github.com/angular/angular-cli/files/2320208/build-lazy-bug.zip)

1) NPM I
2) Open layout.routing.ts
`import { ModuleWithProviders } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { LayoutComponent } from '../components';

const layoutRoutes: Routes = [
    {
        path: '',
        component: LayoutComponent,
        children: [
            {
                path: 'rooms',
                loadChildren: '../../room/room.module#RoomModule'
            }
        ]
    }
];

export const layoutRouting: ModuleWithProviders = RouterModule.forChild(layoutRoutes);
`

and try to run / build application (ng serve), there is an error:
`ERROR in Could not resolve module ../../room/room.module relative to /D/GIT/angular-empty/src/app/layout/layout.module.ts`

JUST resave any file to start rebuild on watch (press ctrl+s):
- no error
- works like a charm


I feel like i have encountered this problem like year ago and it is back.

**Comment by juri33** [comment]
+1

**Comment by glebiuskv** [comment]
+1

**Comment by zhq5101** [comment]
I found the reason about this problem. We need tell compiler the base directory to resolve non-relative module names. The configuration file tsconfig.app.json is the right place to do that.
Open file src/tsconfig.app.json and add a line `"baseUrl": "./",` to indicate the base directory.
This is the final file after I added the new line:
`{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "baseUrl": "./",
    "module": "es2015",
    "types": []
  },
  "exclude": [
    "src/test.ts",
    "**/*.spec.ts"
  ]
}
`

**Comment by filipesilva** [comment]
Thanks for reporting this issue. This issue was originally reported a long time ago and since then we've had many releases, one of which might have addressed this problem. Please update to the most recent Angular CLI version.

If the problem persists after upgrading, please open a new issue, provide a simple repository reproducing the problem, and describe the difference between the expected and current behavior.
